### ISSUE TITLE
Flutter Gradle Plugin uses antipatterns

### ISSUE THREAD

**Issue by bartekpacia** [flutter/flutter#123934]
[Flutter Gradle Plugin](https://github.com/flutter/flutter/blob/3.7.9/packages/flutter_tools/gradle/flutter.gradle) makes use of a few patterns that are considered to be "bad Gradle".

1. Names of AGP tasks are "guessed", instead of hooking up into official extension points.

   I learned about this from [Extend the Android Gradle Plugin article](https://developer.android.com/studio/build/extend-agp). [Here's a paragraph](https://developer.android.com/studio/build/extend-agp#variant-api-artifacts-tasks) that might be useful:

    > When interacting with AGP, use specially made extension points instead of registering the typical Gradle lifecycle callbacks (such as [afterEvaluate()](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#afterEvaluate-org.gradle.api.Action-)) or setting up explicit `Task` dependencies. Tasks created by AGP are considered implementation details and are not exposed as a public API. You must avoid trying to get instances of the Task objects or guessing the `Task` names and adding callbacks or dependencies to those `Task` objects directly.

   This is not what FGP does. Instead, it finds tasks by gluing strings together:

    ```groovy
            Task copyFlutterAssetsTask = project.tasks.create(
                name: "copyFlutterAssets${variant.name.capitalize()}",
                type: Copy,
            ) {
                 // ...
                dependsOn "clean${mergeAssets.name.capitalize()}"
                mergeAssets.mustRunAfter("clean${mergeAssets.name.capitalize()}")
                // ...
            }


            // ...
            def compressAssetsTask = project.tasks.findByName("compress${variant.name.capitalize()}Assets")
            if (compressAssetsTask) {
                compressAssetsTask.dependsOn copyFlutterAssetsTask
            }
    ```

    This is exactly what the Android docs advise against.

2. Dependencies between tasks are defined explicitly, instead of being inferred on the basis of inputs/outputs. Also, FGP tasks output their files into AGP task's output directories.

    I realized this after reading [this great article by a former Gradle engineer](https://melix.github.io/blog/2021/10/gradle-quickie-dependson.html).

    For example, `copyFluterAssets{flavor}` puts its output into the output directory of `merge{flavor}Assets` task [here](https://github.com/flutter/flutter/blob/3.7.6/packages/flutter_tools/gradle/flutter.gradle#L954). The relevant excerpt:

    > multiple tasks contributing to the same output directory is the typical example of what would break caching. In fact, it breaks all kinds of up-to-date checking, that is to say, the ability for Gradle to understand that it doesn’t need to execute a task when nothing changed.

    Let's take the [`copyFlutterAssetsTask`](https://github.com/flutter/flutter/blob/1d90b7654afadaa848aca307971b4cb934f17144/packages/flutter_tools/lib/src/android/gradle_utils.dart#L164-L209) again as an example.

```groovy
Task copyFlutterAssetsTask = project.tasks.create(
    name: "copyFlutterAssets${variant.name.capitalize()}",
    type: Copy,
) {
    dependsOn compileTask
    with compileTask.assets
    if (isUsedAsSubproject) {
        dependsOn packageAssets
        dependsOn cleanPackageAssets
        into packageAssets.outputDir // contributes its own output into output dir of another task
        return
    }
   
   
    dependsOn mergeAssets // "bad"
    dependsOn "clean${mergeAssets.name.capitalize()}"   "bad"
    mergeAssets.mustRunAfter("clean${mergeAssets.name.capitalize()}") forcing order - "bad"
    into mergeAssets.outputDir // contributes its own output into output dir of another task
}
```

It's not anything critical, of course, but I think it's good to have these points noted somewhere and fix them in the long run. I think it'd be good to do it _after_ #122579  because writing Kotlin is faster and more pleasant than writing Groovy :)

This is somewhat related to #119196.

**Comment by bartekpacia** [comment]
Good idea.

As for the lints checks - I think that we should at least make sure that no warnings are emitted by Gradle in a default `flutter create`d app. These warnings often become errors in later Gradle/AGP versions.

**Comment by reidbaker** [comment]
> Good idea.
> 
> Re: Gradle style guide – cool idea, but that'd be quite a lot of effort. For now, if somebody's interested I'd recommend reading:
> 
> * https://github.com/liutikas/gradle-best-practices
> * https://github.com/jjohannes/idiomatic-gradle
> * [Understanding Gradle plugins: the provider API](http://melix.github.io/blog/2022/01/understanding-provider-api.html)
> * [Gradle quickie: laziness](http://melix.github.io/blog/2022/05/gradle-laziness.html)
> * [A Gradle quickie: properly using dependsOn](http://melix.github.io/blog/2021/10/gradle-quickie-dependson.html)
> * [Understanding Gradle playlist](https://youtube.com/playlist?list=PLWQK2ZdV4Yl2k2OmC_gsjDpdIBTN0qqkE)
> 
> Re: lints checks - I think that we should at least make sure that no warnings are emitted by Gradle in a default `flutter create`d app. These warnings often become errors in later Gradle/AGP versions.

Yeah I agree creating a style guide that is solid is a lot of work. My intent was more to indicate that if you run across things that bother you feel free to add a gradle style section so we can start to enforce what you have found to be best practices. 

It may say "Do not do XYZ and instead do ABC" and that is all.

**Comment by SichangHe** [comment]
May be related:

<details>
<summary>Android Studio Build Analyzer report.</summary>

How do I follow the recommendation from Build Analyzer by editing my build.gradle:
:app:copyFlutterAssetsDebug This task frequently determines build duration because of dependencies between its inputs/outputs and other tasks.  Duration: 6.9s / 43.5% Sub-project: :app Plugin: FlutterPlugin Type: org.gradle.api.tasks.Copy Task Execution Categories: Uncategorized  Warnings Consider filing a bug to report this issue to the plugin developer. Generate report
 
Task Setup Issues

This task declares the same output directory as task ':app:mergeDebugAssets': build/app/intermediates/assets/debug As a result, these tasks are not able to take advantage of incremental build optimizations and might need to run with each subsequent build. Learn more 

Recommendation: Edit the plugin(s) to ensure each task specifies a unique output directory.
Reason task ran Output property 'destinationDir' file build/app/intermediates/assets/debug/flutter_assets has been removed. Output property 'destinationDir' file build/app/intermediates/assets/debug/flutter_assets/AssetManifest.bin has been removed. Output property 'destinationDir' file build/app/intermediates/assets/debug/flutter_assets/AssetManifest.json has been removed. 

:app:mergeDebugAssets This task frequently determines build duration because of dependencies between its inputs/outputs and other tasks.  Duration: 5.6s / 35.0% Sub-project: :app Plugin: com.android.application Type: com.android.build.gradle.tasks.MergeSourceSetFolders Task Execution Categories: Uncategorized  Warnings Consider filing a bug to report this issue to the plugin developer. Generate report
 
Task Setup Issues

This task declares the same output directory as task ':app:copyFlutterAssetsDebug': build/app/intermediates/assets/debug As a result, these tasks are not able to take advantage of incremental build optimizations and might need to run with each subsequent build. Learn more 

Recommendation: Edit the plugin(s) to ensure each task specifies a unique output directory.
</details>

Adding this reduces my incremental build time by 6sec:

```groovy
afterEvaluate {
    // Help cache :copyFlutterAssetsDebug.
    cleanMergeDebugAssets.enabled = false
}
```
